src/Core/PS4/Linker.cpp

```diff
@@ -650,37 +669,556 @@ static PS4_SYSV_ABI void ProgramExitFunc() {
+
+        push_abi_regs();
+
+        gen->mov(gen->rax, context_base);
+        gen->mov(gen->qword[gen->rax + host_rsp_offs], gen->rsp);
+
+        gen->mov(rax, context_base);
+
+		// RSP
+		gen->mov(rsp, ptr[rax + rsp.getIdx() * 8]);
+        
+		// pop & store original return address
+        gen->pop(rdx);
+        gen->mov(ptr[rax + trampoline_ret_offs], rdx);
+
+        // args: RDI, RSI, RDX, RCX, R8, R9
+        Reg64 args_64[] = {rdi, rsi, rdx, rcx, r8, r9};
+        for (const auto& reg : args_64) {
+            gen->mov(reg, ptr[rax + reg.getIdx() * 8]);
+		}
+        // args: XMM0, XMM1, XMM2, XMM3, XMM4, XMM5, XMM6 and XMM7
+        for (int i = 0; i < 8; i++) {
+            gen->vmovaps(Ymm(i), ptr[rax + ymm_offs + i * ymm_size]);
+		}
+
+		gen->mov(rax, hle_handler);
+        gen->call(rax) ;// this replaces the original return address
+
+        // rets: RAX, RDX
+        gen->mov(rcx, rax);
+        Reg64 rets_64[] = {rcx /* rax is used as temp */, rdx};
+
+        gen->mov(rax, context_base);
+        gen->mov(ptr[rax + rax.getIdx() * 8], rets_64[0]);
+        gen->mov(ptr[rax + rets_64[1].getIdx() * 8], rets_64[1]);
+        
+        // rets: XMM0
+        gen->vmovaps(ptr[rax + ymm_offs + 0 * ymm_size], Ymm(0));
+
+		// faux ret
+		gen->mov(gen->rax, context_base + 4 * 8);
+        gen->mov(gen->rsp, gen->qword[gen->rax]);
+        gen->pop(gen->rcx);
+        gen->mov(gen->qword[gen->rax], gen->rsp);
+
+        gen->mov(gen->rax, context_base);
+        gen->mov(gen->rsp, gen->qword[gen->rax + host_rsp_offs]);
+        pop_abi_regs();
+
+        gen->mov(gen->rax, ptr[gen->rax + trampoline_ret_offs]);
+        gen->ret();
+    }
+}
+
 static void run_main_entry(u64 addr, EntryParams* params, exit_func_t exit_func) {
-    //reinterpret_cast<entry_func_t>(addr)(params, exit_func); // can't be used, stack has to have a specific layout
+    // reinterpret_cast<entry_func_t>(addr)(params, exit_func); // can't be used, stack has to have a specific layout
 
-	asm volatile (
-        "andq $-16, %%rsp\n"// Align to 16 bytes
-        "subq $8, %%rsp\n"	// videoout_basic expects the stack to be misaligned
+    // Allocate stack for guest thread
+    auto stack_top = 8 * 1024 * 1024 + (u64)VirtualAlloc(0, 8 * 1024 * 1024, MEM_COMMIT, PAGE_READWRITE);
+
+	{
+        auto& rsp = thread_context.gpr[4];
+        auto& rsi = thread_context.gpr[6];
+        auto& rdi = thread_context.gpr[7];
 
-							// Kernel also pushes some more things here during process init
-							// at least: environment, auxv, possibly other things
+        rsp = stack_top;
 
-        "pushq 8(%1)\n"		// copy EntryParams to top of stack like the kernel does
-        "pushq 0(%1)\n"		// OpenOrbis expects to find it there
+        rsp = rsp & ~16;
+        rsp = rsp - 8;
 
-		"movq %1, %%rdi\n"	// also pass params and exit func
-        "movq %2, %%rsi\n"	// as before
+        rsp = rsp - 8;
+        *(void**)rsp = params->argv;
```

Comment:
In this case, it is not a link to the argv array that should be passed, but the array of argv links itself directly. So the size will be variable.

Reply:
Yup, can confirm that the array should be inline there from the assembly
```
TranslateCode: 00000009000B0D18, ctx: 7FF613C59720                                                                      00000009000B0D18  xor rbp, rbp                                                                                          00000009000B0D1B  mov rdi, rsp                                                                                          00000009000B0D1E  lea rsi, [0x000000090010A628]                                                                         00000009000B0D25  and rsp, 0xFFFFFFFFFFFFFFF0                                                                           00000009000B0D29  call 0x00000009000B0D30                                                                               TranslateCode: 00000009000B0D30, ctx: 7FF613C59720                                                                      00000009000B0D30  mov esi, [rdi]                                                                                        00000009000B0D32  lea rdx, [rdi+0x08]                    
```
Funny how that doesn't match with the openorbis source code, must be looking at the wrong place? Hmm.

Reply:
with `long* p` Open Orbis uses ```	char **argv = (void *)(p+1);```, seems I missread that last time. Thanks @red-prig :)